### PR TITLE
add 02Hardware NVIDIA tensorcore1 part

### PR DESCRIPTION
**Pull request by xiaotianzhang01**
增加TensorCore上 文字版和图片

### CODE REVIEW

**Comment by xiaotianzhang01** [comment]
cc @chenzomi12

**Review by chenzomi12 — COMMENTED** [review]
2024.3.8 finshed review

File: 02Hardware/04NVIDIA/image.png
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/image.png:1]: Done

File: 02Hardware/04NVIDIA/images/amp.png
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/images/amp.png:1]: Done

File: 02Hardware/04NVIDIA/images/gpuarh.png
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/images/gpuarh.png:1]: Done

File: 02Hardware/04NVIDIA/01.basic_tc.md [hunk 1]
```diff
@@ -1,6 +1,131 @@
...
+
+在我们深入探讨之前，让我们先简要回顾一下Nvidia GPU架构的演变历程。正如上图所展示的，自2010年推出Fermi架构以来，经过12年的发展，Nvidia GPU已经历了8代架构的迭代，从最初的Fermi架构起，Nvidia的GPU就开始配备CUDA核心。而到了2017年引入的Volta架构，标志着第一代TensorCore核心的诞生。
+
+自从Volta架构搭载了首代TensorCore以来，Nvidia在每一次的架构升级中都不断对TensorCore进行优化和更新，每一轮的更新都带来了新的变化和提升。接下来，我们将逐步深入介绍其原理和这些演进过程。
+
+## 卷积计算
+
+卷积运算是深度学习中最基本也是最关键的操作之一，尤其是在处理图像和视频相关的任务时。而TensorCore则是Nvidia推出的一种专为加速深度学习中的矩阵计算而设计的硬件加速器。要理解卷积与TensorCore之间的关系，我们需要先了解卷积运算的本质。
+
+### CNN vs GEMM
```
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/01.basic_tc.md]: Done

File: 02Hardware/04NVIDIA/01.basic_tc.md [hunk 2]
```diff
@@ -1,6 +1,131 @@
...
+
+在深度学习中，卷积运算通常指的是利用一个小的、可学习的过滤器（或称为卷积核）在输入数据（如图像）上滑动，并在每个位置计算过滤器与其覆盖区域的元素逐点相乘后的总和，这个过程可以捕捉到局部特征。对于多通道输入，卷积运算会对每个通道执行此操作，并将结果累加起来得到最终的输出。当应用于图像处理时，这种机制使得卷积网络能够有效地识别图像中的边缘、纹理等特征。
+
+卷积神经网络CNN一般包含许多卷积层，这些层通过卷积运算提取输入数据的特征。在算法层面上，卷积运算的加速通常涉及到一个关键步骤——数据重排，即执行Im2col操作。
+
+Im2col操作的目的是将卷积运算转换为矩阵乘法，这样做有几个显著的好处。首先，它允许利用已有的高效矩阵乘法算法（如GEMM）来加速卷积计算。其次，这种转换可以减少重复的内存访问，因为在传统的卷积运算中，同一个输入元素可能会被多个卷积核重复使用。
```
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/01.basic_tc.md:36]: Done

File: 02Hardware/04NVIDIA/01.basic_tc.md [hunk 3]
```diff
@@ -1,6 +1,131 @@
...
+
+如上图所示，在CUDA编程体系中，我们并非直接对线程进行控制，也就是图中的弯弯的线，而是通过控制一个Warp，一个Warp包含很多线程（通常为32个线程），这些线程同时并行执行，利用GPU的并行计算能力。
+
+在实际执行过程中，CUDA会对Warp进行同步操作，确保其中的所有线程都达到同步点，并获取相同的数据。然后，这些线程将一起执行矩阵相乘和其他计算操作，通常以16x16的矩阵块为单位进行计算。最终，计算结果将被存储回不同的Warp中，以便后续处理或输出。
+
+我们可以把Warp理解为软件上的一个大的线程概念。它帮助简化了对GPU并行计算资源的管理和利用。通过有效地利用Warp的并行性，CUDA程序可以实现高效、快速的并行计算。
+
+在CUDA程序执行过程中，我们可以通过线程的Warp来调度Tensor Core的执行。多个Tensor Core可以同时通过Warp内的线程来执行计算任务，利用Tensor Core提供的高性能矩阵运算能力。每个Warp内的线程可以利用Tensor Core执行16x16x16的矩阵运算，充分发挥GPU的计算潜能。
+
+![CUDA编程使用TensorCore](images/cuda-program.png)
```
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/01.basic_tc.md]: Done

File: 02Hardware/04NVIDIA/images/cuda-program.png
[the remark this answers is not in the document]
Comment (xiaotianzhang01) [reply 02Hardware/04NVIDIA/images/cuda-program.png:1]: Done

**Comment by xiaotianzhang01** [comment]
cc\  @chenzomi12

**Pull request merged**